File: 0289 Game of Life.java
class Solution {
    private int m = 0, n = 0;
    private int[][] d = {{-1,1},{-1,0},{-1,-1},{0,1},{0,-1},{1,1},{1,0},{1,-1}};
    public void gameOfLife(int[][] board) {        
        m = board.length;
        n = board[0].length;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                check(board, i, j);
            }
        }
        next(board);
    }
    public void check(int[][] board, int i, int j) {
        int cnt = 0;
        for (int k = 0; k < 8; ++k) {
            cnt += alive(board, i + d[k][0], j + d[k][1]);
        }
        if (board[i][j] % 2 == 1) {
            if (cnt == 2 || cnt == 3) board[i][j] += 2;
        } else if (cnt == 3) {
            board[i][j] += 2;
        }
    }
    public int alive(int[][] board, int i, int j) {
        if (i < 0 || j < 0 || i >= m || j >= n) {
            return 0;
        } else {
            return board[i][j] % 2;
        }
    }
    
    public void next(int[][] board) {
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                board[i][j] /= 2;
            }
        }
    }
    
    // public void printboard(int[][] board) {
    //     for (int i = 0; i < m; ++i) {
    //         for (int j = 0; j < n; ++j) {
    //             System.out.print(board[i][j] + " ");
    //         }
    //         System.out.println("");
    //     }
    // }
}

File: 0289 Game of Life.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        # global definitions
        m, n = len(board), len(board[0])
        d = [(i, j) for i in range(-1, 2) for j in range(-1, 2) if j != 0 or i != 0]        
        
        # functions
        def check(board, i, j):
            cnt = 0
            for k in range(8):
                cnt += alive(board, i + d[k][0], j + d[k][1])
            if board[i][j] % 2 == 1:
                if cnt == 2 or cnt == 3: board[i][j] += 2
            elif cnt == 3:
                board[i][j] += 2
                
        def alive(board, i, j):
            if i < 0 or j < 0 or i >= m or j >= n:
                return 0
            else:
                return board[i][j] % 2
        
        def next(board):
            for i in range(m):
                for j in range(n):
                    board[i][j] //= 2      
        
        # check and get next
        for i in range(m):
            for j in range(n):
                check(board, i, j)
        next(board)
